### PR TITLE
Reenable checks and fix errors

### PR DESCRIPTION
**Pull request by benjamin-gorman**
### Description

Fix currently ignored errors and reenable corresponding checks

- `autodoc-typehints`: enable flag (TYPE_CHECKING imports)
- `flake8-bugbear`: reenable checks and fix errors
- `flake8-comprehensions`: reenable checks and fix errors
- `flake8 unused imports`: reenable checks and fix errors
- `flake8-mypy`: remove checks (no longer using)
- `mypy-torch`: reenable checks and fix errors

### Status

**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: monai/apps/datasets.py [hunk 1]
```diff
@@ -58,7 +61,7 @@ def __init__(
         self,
         root_dir: str,
         section: str,
-        transform: Callable[..., Any] = LoadPNGd("image"),
+        transform: Callable = DEFAULT_LOADPNGD,
```
Comment (benjamin-gorman) [inline monai/apps/datasets.py:64]: > B008 Do not perform function calls in argument defaults. The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call. If this is intended, assign the function call to a module-level variable and use that variable as a default value.

Is this intended or should I do something like

```python
transform: Optional[Callable] = None

if transform is None:
    transform = LoadPNGd("image")
```

File: research/lamp-automated-model-parallelism/data_utils.py [hunk 2]
```diff
@@ -53,7 +53,7 @@ def load_data_and_mask(data, mask_data):
     img = np.load(data)  # z y x
     img = pad_xform(img[None])[0]
     item = dict(image=img, label=[])
-    for idx, maskfnm in enumerate(mask_data):
+    for _, maskfnm in enumerate(mask_data):
```
Comment (benjamin-gorman) [inline research/lamp-automated-model-parallelism/data_utils.py:56]: > B007 Loop control variable '{}' not used within the loop body. If this is intended, start the name with an underscore.

Should I rewrite the loop if it has redundancies or is it better to leave it

```python
for maskfnm in mask_data:
```

File: monai/apps/datasets.py (hunk 1 above)
Comment (wyli) [reply monai/apps/datasets.py:64]: the proposed change of the default behaviour will actually break one of the tutorials. looks like we need a solution to support both no transform and default_loading @Nic-Ma

File: research/lamp-automated-model-parallelism/data_utils.py (hunk 2 above)
Comment (wyli) [reply research/lamp-automated-model-parallelism/data_utils.py:56]: yes please change this, thanks!

File: monai/apps/datasets.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/apps/datasets.py:64]: Hi @wyli and @benjamin-gorman ,

1. I think we can use `transform: Optional[Callable] = DEFAULT_LOADPNGD` to support both `default transform` and `None` transform.
2. And then I will change the `CacheDataset` and `PersistentDataset` to support no transforms(actually, there is a bug in `PersistentDataset` that it has default value `None` for transform, but can't work with `None`).

What do you think?

Thanks.

Comment (wyli) [reply monai/apps/datasets.py:64]: @Nic-Ma the problem here is that `DEFAULT_LOADPNGD` is a single globally shared instance, may have some surprising side effects. @benjamin-gorman I think we can keep it for now, add `# ignore`if needed. it is an issue and needs another PR to address it

Comment (Nic-Ma) [reply monai/apps/datasets.py:64]: @wyli , yes, I think maybe we should not use the global variable for it now.
@benjamin-gorman , I am OK with both solutions:
1. `transform: Optional[Callable] = DEFAULT_LOADPNGD` with clear description for the potential issue.
2. `transform: Optional[Callable] = LoadPNGd("image")`.

Thanks.

Comment (benjamin-gorman) [reply monai/apps/datasets.py]: Reverted the changes and made issue #727

File: monai/handlers/mean_dice.py [hunk 3]
```diff
@@ -72,13 +72,13 @@ def reset(self) -> None:
...
     @reinit__is_reduced
-    def update(self, output: Sequence[Union[torch.Tensor, dict]]) -> None:
+    def update(self, output: Sequence[torch.Tensor]) -> None:
         if not len(output) == 2:
             raise ValueError("MeanDice metric can only support y_pred and y.")
         y_pred, y = output
         score = self.dice(y_pred, y)
         assert self.dice.not_nans is not None
-        not_nans = self.dice.not_nans.item()
+        not_nans = cast(int, self.dice.not_nans.item())
```
Comment (wyli) [inline monai/handlers/mean_dice.py:81]: this could be the actual casting `int(self.dice.not_nans.item())`

Comment (benjamin-gorman) [reply monai/handlers/mean_dice.py]: changed

**Comment by benjamin-gorman** [comment]
> thanks, looks like we can close ticket #476 after this PR?

There are still various missing type hints; I think we will need to resolve them first.

**Pull request merged**